actor-share-conversion/src/gf2_128/sender.rs

```diff
@@ -0,0 +1,229 @@
+    }
+}
+
+#[async_trait]
+impl<T> SendTape for SenderControl<T>
+where
+    T: Handler<SendTapeMessage, Return = Result<(), ShareConversionError>>,
+{
+    async fn send_tape(self) -> Result<(), ShareConversionError> {
+        self.0
+            .send(SendTapeMessage)
+            .await
+            .map_err(|err| ShareConversionError::Other(err.to_string()))?
+    }
+}
+
+#[async_trait]
+impl<T, OT, U, V, W> Handler<M2AMessage<Vec<u128>>> for Sender<T, OT, U, V, W>
+where
+    T: AsyncFactory<OT, Config = OTSenderConfig, Error = OTFactoryError> + Send + 'static,
+    OT: ObliviousSend<[Block; 2]> + Send + 'static,
+    U: Gf2_128ShareConvert + Send + 'static,
+    V: MuxChannelControl<Gf2ConversionMessage> + Send + 'static,
+    W: Recorder<U> + Send + 'static,
+    IOSender<T, OT, U, W>: MultiplicativeToAdditive<FieldElement = u128>,
+{
+    type Return = Result<Vec<u128>, ShareConversionError>;
+
+    async fn handle(
+        &mut self,
+        message: M2AMessage<Vec<u128>>,
+        _ctx: &mut Context<Self>,
+    ) -> Self::Return {
+        match self.state {
+            State::Setup(ref mut state) => state.m_to_a(message.0).await,
+            _ => Err(ShareConversionError::Other(String::from(
+                "Actor is not setup",
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl<T, OT, U, V, W> Handler<A2MMessage<Vec<u128>>> for Sender<T, OT, U, V, W>
+where
+    T: AsyncFactory<OT, Config = OTSenderConfig, Error = OTFactoryError> + Send + 'static,
+    OT: ObliviousSend<[Block; 2]> + Send + 'static,
+    U: Gf2_128ShareConvert + Send + 'static,
+    V: MuxChannelControl<Gf2ConversionMessage> + Send + 'static,
+    W: Recorder<U> + Send + 'static,
+    IOSender<T, OT, U, W>: AdditiveToMultiplicative<FieldElement = u128>,
+{
+    type Return = Result<Vec<u128>, ShareConversionError>;
+
+    async fn handle(
+        &mut self,
+        message: A2MMessage<Vec<u128>>,
+        _ctx: &mut Context<Self>,
+    ) -> Self::Return {
+        match self.state {
+            State::Setup(ref mut state) => state.a_to_m(message.0).await,
+            _ => Err(ShareConversionError::Other(String::from(
+                "Actor is not setup",
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl<T, OT, U, V> Handler<SendTapeMessage> for Sender<T, OT, U, V, Tape>
+where
+    T: AsyncFactory<OT, Config = OTSenderConfig, Error = OTFactoryError> + Send + 'static,
+    OT: ObliviousSend<[Block; 2]> + Send + 'static,
+    V: MuxChannelControl<Gf2ConversionMessage> + Send + 'static,
+    U: Gf2_128ShareConvert + Send + 'static,
+    IOSender<T, OT, U, Tape>: SendTape,
+{
+    type Return = Result<(), ShareConversionError>;
+
+    async fn handle(&mut self, _message: SendTapeMessage, ctx: &mut Context<Self>) -> Self::Return {
+        let state = std::mem::replace(&mut self.state, State::Complete);
+        let State::Setup(state) = state else {
+            return Err(ShareConversionError::Other(String::from(
```

Comment:
here, unlike in setup() self is not returned into the previous state. i think it should be reverted to the previous state if this is not a critical error.

Reply:
sry, i meant that it should be reverted just before the line
`return Err(ShareConversionError::Other(String::from(
                "Actor is not setup"
            )));
`
not after `state.send_tape().await`

If it was up to me, I would make it so that all errors when the handler is called in the wrong state are critical errors and would shut down the Actor. 
But since you chose a less paranoid approach, we should revert the state when a non-critical error happens.